### DISCUSSION TITLE
ESP8266 Uploading micropython firmware ok but not working

### DISCUSSION THREAD

**Discussion by fudi500** [micropython/micropython#10974] [ESP8266]
Model : ESP8266 NodeMcu V3
Auto-detected Flash size: 4MB

Hello, I'm starting to play with microphyton on esp8266, I install everything from different guides but always the same effect. I used different versions of microphytin and different ways of uploading the firmware.  The microphyton upload is successful, but Thony, uPyCraft_V1.1, or PuTTY detect the device on the COM port but behave as if microhyton is not installed.

Maybe someone faced this problem?

```
C:\Users\fudit>esptool.py --port COM4 erase_flash
esptool.py v4.5.1
Serial port COM4
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: b4:e6:2d:55:2c:76
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 12.5s
Hard resetting via RTS pin...

C:\Users\fudit>esptool.py --port COM4 write_flash --flash_size=detect -fm dio 0x00000 E:\esp8266-512k-20220618-v1.19.1.b
in
esptool.py v4.5.1
Serial port COM4
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: b4:e6:2d:55:2c:76
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash will be erased from 0x00000000 to 0x00078fff...
Flash params set to 0x0240
Compressed 495084 bytes to 331639...
Wrote 495084 bytes (331639 compressed) at 0x00000000 in 29.2 seconds (effective 135.5 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

C:\Users\fudit>
```
Afther that 
Thonny 
```
Device is busy or does not respond. Your options:

  - wait until it completes current work;
  - use Ctrl+C to interrupt current work;
  - reset the device and try again;
  - check connection properties;
  - make sure the device has suitable MicroPython / CircuitPython / firmware;
  - make sure the device is not in bootloader mode.
 ```

**Comment by sethstenzel** [comment]
So when you connect to the device directly with [Putty ](https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html)and then hit the reset button on the ESP and do you see any information coming through at all?

**Reply by sethstenzel** [reply]
Also maybe try decreasing the file system size yourself as a test:
`esptool.py --port COM4 write_flash --flash_mode dio --flash_size 1MB 0x0 E:\esp8266-512k-20220618-v1.19.1.b`

**Reply by 21dczhang** [reply]
I faced the same problem，and I use the command what you said， but also can't solve the problem. This is Putty‘s  reply.
{$$▒▒|▒l▒|
▒l▒
b|▒▒▒{▒b▒#▒▒nn▒$oo▒▒▒
bp▒▒l{${$p▒o▒
▒$
#o▒|▒
▒r▒b▒▒nn▒$▒▒l`▒n
▒lnr▒ےo▒▒▒`p▒o▒
#o▒|
b▒▒oo▒
▒l ▒on
l`o{▒▒▒n
▒▒  ▒nl▒▒▒#▒n
▒▒oo▒
▒{lp▒o▒
#o▒▒n
#▒▒on▒l▒$ ▒o$ nr▒▒▒o
▒▒ r▒▒n
▒▒ {ۓo$▒|▒▒`r▒n▒#▒n#p
Do you have any ideas?

**Reply by massimosala** [reply]
This could be a baudrate mismatch.  On your terminal app, set the baud rate to 115200.
Note: even with the right settings, Teraterm (I use this program on Windows) sometimes displays gibberish characters.
I simply close and re-open Teraterm.

**Comment by robert-hh** [comment]
Can you show a picture of the board? 
You can as well try to flash with different flash modes, Option --flash_mode of esptool. Often `--flash_mode DOUT` works better.

**Reply by lechulech189** [reply]
I had the same problem as stated initially and this solution worked! Thank you!

**Comment by 21dczhang** [comment]
oh，I meet the same problem，have any solution？

**Reply by 21dczhang** [reply]
I think I find the solution，just change the flash size when you burning firmware，just like this
** esptool.py --port COM4 write_flash --flash_mode dio --flash_size 4MB 0x0 C:\Users\Aurora\Desktop\esp8266-20220618-v1.19.1.bin**

**Reply by renshaozhang** [reply]
thanks a lot ,i use the nodeMCU ,met the same problem, now it can work .

**Reply by harribg-git** [reply]
Thank you sooo much!!! Great job! As the others here I’ve tried all the ideas found, but nothing worked. This was it!

**Comment by dnguyen76** [comment]
Hello
   II have the same problem with 2 cards on a set of 3
The 3rd one works perfectly.
These cards are clones of the WEMOS D1 mini  with a marking D1 m1 n1 V3.0.0  with 4MB and ESP8266-EX chip
I think it is a bad batch of manufacturing 

![WEMOS D1 Clone](https://user-images.githubusercontent.com/25987158/226692777-5fe7e780-dfda-4b3e-9a0c-1e4be80b74fe.jpg)

**Comment by dnguyen76** [comment]
The previous tests  has been made with Thonny
If i use esptool alone i succeed to get micropython interpreter !!

The esptool command is 👍 
------------------------------
esptool.py --port COM27 write_flash --flash_size=detect -fm dio 0x00000 esp8266-20220618-v1.19.1.bin
esptool.py v3.0
Serial port COM27
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: a8:48:fa:ce:2d:ee
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0240
Compressed 634844 bytes to 419808...
Wrote 634844 bytes (419808 compressed) at 0x00000000 in 34.2 seconds (effective 68.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

**Reply by robert-hh** [reply]
The usual lesson: If it fails with Thonny, try without Thonny.

**Reply by dnguyen76** [reply]
I looked in detail at the esptool command that works and I noted the parameter -fm dio

In fact **in Thonny you have to select Dual I/O (dio) as flash mode : It works**
![Thonny firmware installer](https://user-images.githubusercontent.com/25987158/227303656-30291ce8-6258-412f-8881-c9698a857576.jpg)

**Reply by dnguyen76** [reply]
robert-hh    did you try Dual I/O (dio) to flash firmware with Thonny ?

**Reply by robert-hh** [reply]
I never use Thonny. Too much "magic" under the hood.

**Comment by robert-hh** [comment]
DIO is the default mode of esptool.py. If uploaded firmware using edptool.py fails, either one has to select a different flash mode, like DOUT, or specify the flash size instead of using detect for the flash size.

**Reply by dnguyen76** [reply]
which tool are you using to develop micropython code ?

**Reply by robert-hh** [reply]
Just simple ones: Code editor either on the PC, like VSC or Sublime text, or on the board, the self written pye editor. For script transfer I use what's available, like recently mpremote, earlier ampy or rshell, or ftp if feasible. Firmware upload is done by esptool.py, either by directly calling it or as part of the ESP8266 Makefile. And obviously as the most primitive tools head & hand.

**Comment by slash128v6** [comment]
I have been having similar issues with several ESP8266 NodeMCU dev boards. They flash fine but then they constantly reset. I have tried several cables, firmware versions, flash command variations and even different laptops. I have some ESP32 boards that are working great, just the ESP8266 flavors having trouble.

>>> 
 ets Jan  8 2013,rst cause:4, boot mode:(3,6)

wdt reset
load 0x40100000, len 30720, room 16 
tail 0
chksum 0xba
load 0x3ffe8000, len 996, room 8 
tail 12
chksum 0x2c
ho 0 tail 12 room 4
load 0x3ffe83f0, len 1080, room 12 
tail 12
chksum 0x43
csum 0x43
����o�|2{��o|�llld`#��|{�$�N��o�l ��r�$���$`��{�l�l�$`��r�p����$l`rl��c
MicroPython v1.19.1 on 2022-06-18; ESP module with ESP8266
Type "help()" for more information.
>>>

**Reply by lgcwzs** [reply]
I have the same problem with 2 cards.   Try to  esptool, finally ok： esptool.py --port COM27 write_flash --flash_size=detect -fm dio 0x00000 esp8266-20220618-v1.19.1.bin

**Reply by robert-hh** [reply]
Did you try the various `-fm` options. A common alternative option would be `-fm dout`. 

Besides that, you could use a more recent firmware version. But that is not related to the problem you see.

**Comment by amorawala** [comment]
Hello. I have flashed uPython on my ESP-01 with 1MBytes flash. My board is home made, with RS232C interface (for the other projects I work with) and I use FTDI USB to RS232C interface board. All hardware is confirmed working because I have used this to flash lua.init, AT commands and more and they all work. 

There is no REPL boot prompt. This is my erase flash command:
With GPIO0 grounded:
C:\Users\User>esptool.py --chip esp8266 erase_flash
esptool.py v4.5.1
Found 1 serial ports
Serial port COM4
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: bc:dd:c2:b3:10:74
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 2.7s
Hard resetting via RTS pin...

And the command to flash on ESP-01:
Flash with the command (keep GPIO0 grounded): 
Command typed, with full path of .bin file (since .bin is not in PATH)
C:\Users\User>esptool.py --chip esp8266 --port COM4 write_flash -z 0x1000 F:\Applications_Software\Python-Micro\esp8266-1m-20220618-v1.19.1.bin
esptool.py v4.5.1
Serial port COM4
Connecting...
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: bc:dd:c2:b3:10:74
Stub is already running. No upload is necessary.
Configuring flash size...
Flash will be erased from 0x00001000 to 0x00091fff...
Compressed 591476 bytes to 392426...
Wrote 591476 bytes (392426 compressed) at 0x00001000 in 35.0 seconds (effective 135.3 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

When connected to Thonny, the message is that the device is busy....the standard error.
When connected to PUTTY at 115200 baud, no flow control, I get junk on the screen. When I set the baud to 74480, I get the boot error message: 
ets Jan  8 2013,rst cause:2, boot mode:(3,7)
ets_main.c

After going through many web sites and web pages over the last one week, I have nt been able to resolve this issue. I am keen to get uPy working on my ESP. but am stuck with this problem. 
Can someone please assist?

**Reply by amorawala** [reply]
Hello. After much trials, the following commands offered the solution:
C:\Users\User>esptool.py --chip esp8266 --port COM3 --baud 115200 write_flash --flash_mode dout --flash_freq 26m --flash_size 1MB 0x0 F:\Applications_Software\Python-Micro\esp8266-1m-20220618-v1.19.1.bin

I specified flash mode DOUT, flash freq 26m (because the ESP-01 is clocked at 26MHz), and size 1MB. 

I got the response properly. The on a terminal with 74480 baud, i saw the following:
ets Jan  8 2013,rst cause:2, boot mode:(3,6)	

load 0x40100000, len 30720, room 16
tail 0
chksum 0x90
load 0x3ffe8000, len 996, room 8
tail 12
chksum 0x80
ho 0 tail 12 room 4
load 0x3ffe83f0, len 1080, room 12
tail 12
chksum 0xf5
csum 0xf5
rf cal sector: 251
freq trace enable 0
rf[112] : 00
rf[113] : 00
rf[114] : 01

SDK ver: 2.2.0-dev(9422289) compiled @ Nov  3 2017 19:40:08
phy ver: 1136_0, pp ver: 10.2

And on the terminal at 115200 baud, on bootup:
-A-5ûQ-,5`BAP(bekQ®hb@O%²Õ=!V*iLp[`~
                                    ~i
                                      é«ùÍIny
MicroPython v1.19.1 on 2022-06-18; ESP module (1M) with ESP8266
Type "help()" for more information.
>>> 
REPL prompt received. I tried print("hello"). it worked.
So I specified all parameters I thought may be wrongly put in the .bin image file. Especially the flash mode, frequency and size, since the esp8266 devices may vary in these parameters. Please check your own devices and specify these parameters.
Hope this helps.

**Comment by bikingfreak** [comment]
> Model : ESP8266 NodeMcu V3 Auto-detected Flash size: 4MB
> 
> Hello, I'm starting to play with microphyton on esp8266, I install everything from different guides but always the same effect. I used different versions of microphytin and different ways of uploading the firmware. The microphyton upload is successful, but Thony, uPyCraft_V1.1, or PuTTY detect the device on the COM port but behave as if microhyton is not installed.
> 
> Maybe someone faced this problem?
> 
> ```
> C:\Users\fudit>esptool.py --port COM4 erase_flash
> esptool.py v4.5.1
> Serial port COM4
> Connecting....
> Detecting chip type... Unsupported detection protocol, switching and trying again...
> Connecting....
> Detecting chip type... ESP8266
> Chip is ESP8266EX
> Features: WiFi
> Crystal is 26MHz
> MAC: b4:e6:2d:55:2c:76
> Uploading stub...
> Running stub...
> Stub running...
> Erasing flash (this may take a while)...
> Chip erase completed successfully in 12.5s
> Hard resetting via RTS pin...
> 
> C:\Users\fudit>esptool.py --port COM4 write_flash --flash_size=detect -fm dio 0x00000 E:\esp8266-512k-20220618-v1.19.1.b
> in
> esptool.py v4.5.1
> Serial port COM4
> Connecting....
> Detecting chip type... Unsupported detection protocol, switching and trying again...
> Connecting....
> Detecting chip type... ESP8266
> Chip is ESP8266EX
> Features: WiFi
> Crystal is 26MHz
> MAC: b4:e6:2d:55:2c:76
> Uploading stub...
> Running stub...
> Stub running...
> Configuring flash size...
> Auto-detected Flash size: 4MB
> Flash will be erased from 0x00000000 to 0x00078fff...
> Flash params set to 0x0240
> Compressed 495084 bytes to 331639...
> Wrote 495084 bytes (331639 compressed) at 0x00000000 in 29.2 seconds (effective 135.5 kbit/s)...
> Hash of data verified.
> 
> Leaving...
> Hard resetting via RTS pin...
> 
> C:\Users\fudit>
> ```
> 
> Afther that Thonny
> 
> ```
> Device is busy or does not respond. Your options:
> 
>   - wait until it completes current work;
>   - use Ctrl+C to interrupt current work;
>   - reset the device and try again;
>   - check connection properties;
>   - make sure the device has suitable MicroPython / CircuitPython / firmware;
>   - make sure the device is not in bootloader mode.
> ```

I am having the same issue with 3 Espressif ESP8266-01 boards: I can flash microPython to the boards but in none of the methods of downloading microPython code modules (Thonny, picocom,ampy and repl) was I able to download even a small boot.py module. The errors range from error 1 in Thonny to ampy.pyboard.
PyboardError: ('exception', b'', b'Traceback (most recent call last):\r\n  File "<stdin>", line 1, in <module>\r\nOSError: [Errno 19] ENODEV\r\n') 

The port_diag command I ran at the microPython prompt gave:
import port_diag
FlashROM:
Flash ID: 1460c4 (Vendor: c4 Device: 6014)
Flash bootloader data:
Byte @2: 00
Byte @3: 20 (Flash size: 1MB Flash freq: 40MHZ)
Firmware checksum:
size: 634000
md5: 4796b17c5b672462934dd33d0e4a1ce7
True

Networking:
STA ifconfig: ('192.168.1.26', '255.255.255.0', '192.168.1.1', '192.168.1.1')
AP ifconfig: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '192.168.1.1')
Free WiFi driver buffers of type:
0: 8 (1,2 TX)
1: 0 (4 Mngmt TX(len: 0x41-0x100))
2: 8 (5 Mngmt TX (len: 0-0x40))
3: 4 (7)
4: 7 (8 RX)
lwIP PCBs:
Active PCB states:
Listen PCB states:
TIME-WAIT PCB states:

So this begs the question: does  the 8266-01 have enough flash memory to run microPython code. BTW, if I typed in a few commands, like 
from machine import Pin
>>> Pin(2, Pin.OUT).value(1)
>>> Pin(2, Pin.OUT).value(0)
>>> Pin(2, Pin.OUT).value(1)
it worked. Also copying and pasting the contents of my boot.py also worked but I could not download the file over the serial adapter from my desktop.

**Reply by bikingfreak** [reply]
Oh, and ESPHome just hangs.

**Reply by scruss** [reply]
> does the 8266-01 have enough flash memory to run microPython

Most ESP-01 modules have 1 MB of Flash or less. There are cut-down MicroPython installations for those boards: [MicroPython - Python for microcontrollers](https://micropython.org/download/ESP8266_GENERIC/)

**Comment by amorawala** [comment]
Hello Kevin.
I had a similar problem. After some searching on the Web, this is what I
did:

*ESP-07 – 4MiB        -- this is my ESP module*

Erase flash (GPIO grounded, baud 115200, COM3, cmd terminal) - first erase
flash

C:\Users\User>esptool.py --chip esp8266 erase_flash         (Enter)

esptool.py v4.5.1

Found 1 serial ports

Serial port COM3

Connecting....

Chip is ESP8266EX

Features: WiFi

Crystal is 26MHz

MAC: 18:fe:34:a2:73:66

Uploading stub...

Running stub...

Stub running...

Erasing flash (this may take a while)...

Chip erase completed successfully in 10.9s

Hard resetting via RTS pin...

Flashing version v1.20.0 with 2+ MiB firmware (620KB)
After erasing flash use this command to flash micropython. Use <flash_mode
dout>, <specify flash_freq, flash_size>, give full path of your firmware. I
used the firmware <esp8266-20230426-v1.20.0.bin>

C:\Users\User>esptool.py --chip esp8266 --port COM3 --baud 115200
write_flash --flash_mode dout --flash_freq 26m --flash_size 4MB 0x0
F:\Applications_Software\Python-Micro\esp8266-20230426-v1.20.0.bin

esptool.py v4.5.1

Serial port COM3

Connecting...

Chip is ESP8266EX

Features: WiFi

Crystal is 26MHz

MAC: 18:fe:34:a2:73:66

Stub is already running. No upload is necessary.

Configuring flash size...

Flash will be erased from 0x00000000 to 0x0009afff...

Flash params set to 0x0341

Compressed 634016 bytes to 420365...

Wrote 634016 bytes (420365 compressed) at 0x00000000 in 37.5 seconds
(effective 135.3 kbit/s)...

Hash of data verified.



Leaving...

Hard resetting via RTS pin...

Checking flash memory:

C:\Users\User>esptool.py flash_id

esptool.py v4.5.1

Found 1 serial ports

Serial port COM4

Connecting....

Detecting chip type... Unsupported detection protocol, switching and trying
again...

Connecting...

Detecting chip type... ESP8266

Chip is ESP8266EX

Features: WiFi

Crystal is 26MHz

MAC: 18:fe:34:a2:73:66

Uploading stub...

Running stub...

Stub running...

Manufacturer: ef

Device: 4016

Detected flash size: 4MB

Hard resetting via RTS pin...

On Thonny IDE, REPL prompt after message (remove GPIO0 short to ground):

*MicroPython v1.20.0 on 2023-04-26; ESP module with ESP8266*
*Type "help()" for more information.*



Hope this helps.

**Comment by bikingfreak** [comment]
Thanks for all the help. All the preceding worked (but one 8266 seemed to be bad (?), until I got to Thonny and all I can get there is : Process ended with exit code 1. I can execute Python commands at the prompt and even copy, paste and run code but cannot save the modules. Could it be I don't have enough memory on that device? It's only an 8266-01.

I am trying to load all the Python code I am currently running on 5 ESP32s but cannot progress beyond this point. I am also trying ampy, ESPHome and Arduino and will try those again.

**Comment by bikingfreak** [comment]
ampy and ESPHome  just hang. WIth Arduino, after uploading a simple sketch (WiFi scan) and resetting I get:
File "_boot.py", line 13, in <module>
  File "inisetup.py", line 44, in setup
  File "inisetup.py", line 16, in check_bootsec
  File "flashbdev.py", line 13, in readblocks
OSError: [Errno 5] EIO

MicroPython v1.20.0 on 2023-04-26; ESP module with ESP8266
Type "help()" for more information.
>>> 
So I will now erase the flash memory and redo that. I successfully created configured ESPHome and uploaded it to the 8266.

**Comment by amorawala** [comment]
Hi Kevin.
ESP-01, by default has less than 1MiB flash. We need 1MiB or more to load
Python. I have 2 units of ESP-01 and replaced the flash with 1 MiB. That
works well.
You can buy the flash memory and replace the original ones.
Regards

**Comment by bikingfreak** [comment]
As you can see from this, I indeed have 1 Mb flash and is that enough for microPython, which I load with Thonny, and 50k of Python code? It seems strange that I have enough to download 634 kb of microPython and it croaks when I try to store even boot.py, which is under 1 kb.
esptool.py flash_id
esptool.py v4.6.2
Loaded custom configuration from /home/kevin/Orange_Pi/esptool.cfg
Found 1 serial ports
Serial port /dev/ttyUSB0
Connecting...
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting...
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 68:c6:3a:fc:0e:67
Stub is already running. No upload is necessary.
Manufacturer: c4
Device: 6014
Detected flash size: 1MB
Hard resetting via RTS pin...

**Comment by bikingfreak** [comment]
The device now shows up on Home Assistant as an ESPHome device via WiFi. I have not configured any pins or actions yet so I will attack that now, though the 8266 is a much more primitive experience than the ESP32. It would be very convenient to have a breakout board rather than jumpers and a breadboard, but that's an issue for a separate topic.

Thanks for your help.

**Comment by robert-hh** [comment]
The 1M devices should have a file system of about 400k size. I suggest to use mpremote or s simple terminal emulator like Putty for REPL or when trying to run devices. Thonny does quite a few things under the hood, which may fail, and then you do not know whether it is a problem of the device or of Thonny.

**Reply by bikingfreak** [reply]
Robert, thanks. I will proceed with using ESPHome since that is the first and only way I have been able to get more than a microPython prompt. Neither  Thonny, REPL nor ampy nor picocom worked for me.

**Comment by subratred** [comment]
Latest update using Thonny: For CH340 NodeMCU v3 ESP8266

1. Use 0x address
2. use dio
3. 11500 baud rate
4. 4MiB flash

**Reply by DizzyFrink** [reply]
Thank you for the message. Works for me too.

**Comment by SteKme** [comment]
Same pain of garbled mess in the Thonny's console, surprising resolution: this GPIO0 pin used to flash the device: 

in my case, I grounded it for flashing and kept during flashing (used DIO in thonny's esptool flash options). Then physically disconnected.

Then connected IO0 to 3.3V via 10k resistor and booted. Got flood of mess in the Thonny's console.
**Then physically disconnected IO0 and left completely floating. Now surprisingly works fine?!!**

I kind of suspect some loop in my case: I have a tiny board, connected UART over one USB>UART board (to ESP, I connect TX,RX and GND from this); also, on another USB port, I have RPi pico and get 3.3V+GND from that. That means 2 USB ports possibly interfering..

**Comment by bikingfreak** [comment]
Thanks for your information. I gave up on 8266s completely - too much hassle for the small difference in price compared with ESP32s.

**Reply by SteKme** [reply]
I need it for its u.FL connector. I use PiPicos normally, but this allows me to connect a higher gain external wifi antenna, that's why: however it is such a pain, indeed... with pico, I just connect usb connector and it works, but this tiny ESP hasn't it.

**Reply by SteKme** [reply]
Let me update, how I am able to write firmware to ESP with ease now - for record for anyone interested, hopefully it may prove useful...

Either ESP8266 with the onboard wifi antenna, or the one with uFL connector.

- Enable pin - via 10kohm to +3.3V
- same RST pin
- GPIO0 via 10k to GND for flashing, othervise to 3.3V for normal operation -- for practical purposes, use jumper here? After ESP boots up, can be used for GPIO out.
- GPIO15 via 10k to GND. (possibly - after boot is over, can be used for GPIO out? but I keep it grounded)
- GPIO2 must be pulled high during boot. Can be used for GPIO out after boot. (E.g. You can use +3.3V>>LED>>GPIO2 - during boot, this pin is pulled high, and then setting the pin low from uPy makes the led shine.)

Have prolific USB>UART convertor with 5/3.3V support (hopefully all have this), set 3.3V voltage there with a jumper. Connect its gnd, 3.3v and both uarts to ESP's pins
Firmware size - even cheapest ESPs I had at hand have plenty of flash space. Just RAM is so limited! I used a normal ESP firmware. 

Thonny - menu run, config interpreter dialog, select ESP8266 and select COM port. Bottom right in the dialog, button Install micropy over esptool. Next dialog opens. Select COM port, uPy family is ESP8266, variant local file, target address 0x0, install speed 115200, flash mode dio, flash size detect. Checkbox disable stub loader not checked. Bottom right of this dialog - select local file, point to downloaded .bin file. Press install.
To detect ESPs flash size - bottom right, there is a feature detection button, let it detect.

In my experience - ALL pins must be pulled high and low as described above or else ESP will boot into the messy state in which the console is full of the erratic output, however if wires are correctly set up, it's fine. At this moment, I am able to use ESP as easily and reliably as RP2 that has USB connector onboard and again, I use ESP for wifi and its uFL connector to connect a higher gain wifi antenna, or where physical space is a limitation. Its about 80k of available RAM after uPy boots is a dramatic limitation, indeed.

**Comment by dnguyen76** [comment]
I have written 8266 micropython interpreter with esptool in a Wemos D1 mini
Chip is ESP8266EX Features: WiFi Crystal is 26MHz
Configuring flash size... Auto-detected Flash size: 4MB

It's working with well with  esp8266_GENERIC-20220618-v1.19.1.bin

But with last version ESP32_GENERIC-20241129-v1.24.1.bin seems not working  esp8266 is busy and i got these 3 esptool messages
WARNING: Suspicious segment 0xee, length 0
WARNING: Suspicious segment 0x18f00, length 16777216
Warning: Image file at 0x0 is not a valid ESP8266 image, so not changing any flash settings.

Is it a bug for this version ESP32_GENERIC-20241129-v1.24.1.  ?

**Reply by robert-hh** [reply]
I cannot replicate that behavior with a Wemos D1. Try to start over:
- download the Image file again
- do a full flash erase with `esptool.py --port /dev/ttyUSB0 --baud 460800 erase_flash`
- reload the firmware.

If that still fails, there may be a hardware problem or a problem clean downloading the file. For verification:
```
md5sum ESP8266_GENERIC-20241129-v1.24.1.bin
cbb56a027a4892507faae602669c850c  ESP8266_GENERIC-20241129-v1.24.1.bin
```

**Reply by dnguyen76** [reply]
I have checked md5sum and get also cbb56a027a4892507faae602669c850c
i  wrote again flash after erase it but also same problem
Size is Compressed 1691664 bytes to 1109554 larger than esp8266_GENERIC-20220618-v1.19.1.bin Compressed 634844 bytes to 419808...
So perhaps a failing board for some part of flash memory

**Reply by robert-hh** [reply]
The size difference is too large for the version change 1.19.1 to 1.24.1. I'm surprised about the file sizes you mention. Here I have:
```
634844  ESP8266_GENERIC-20220618-v1.19.1.bin
635956  ESP8266_GENERIC-20241129-v1.24.1.bin
636156  ESP8266_GENERIC-20250228-v1.25.0-preview.336.g69ffd2aaf.bin
495084  ESP8266_GENERIC-FLASH_512K-20220618-v1.19.1.bin
500280  ESP8266_GENERIC-FLASH_512K-20241129-v1.24.1.bin
```
Did you really get the download images from https://micropython.org/download/ESP8266_GENERIC/

**Reply by dnguyen76** [reply]
You are right , i had mixed **inadvertently** 8266 and **ESP32** firmware ( ESP32_GENERIC-20241129-v1.24.1.bin) 
Sorry and thanks for your help
